feat: open prompt in configured external editor

sayan-oai opened the pull request:
Add `ctrl+g` shortcut to enable opening current prompt in configured editor (`$VISUAL` or `$EDITOR`).


- Prompt is updated with editor's content upon editor close.
- Paste placeholders are automatically expanded when opening the external editor, and are not "recompressed" on close
  - They could be preserved in the editor, but it would be hard to prevent the user from modifying the placeholder text directly, which would drop the mapping to the `pending_paste` value
- Image placeholders stay as-is
- `ctrl+g` explanation added to shortcuts menu, snapshot tests updated


https://github.com/user-attachments/assets/4ee05c81-fa49-4e99-8b07-fc9eef0bbfce

sayan-oai committed:
enable opening prompt in /

sayan-oai committed:
expand pending_pastes when opening editor

sayan-oai committed:
add shortcut to footer and improve errors

sayan-oai commented:
@codex review this

nornagon-openai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1004,6 +1011,70 @@ impl App {
+        let was_alt_screen = tui.is_alt_screen_active();
+        if was_alt_screen {
+            let _ = tui.leave_alt_screen();
+        }
```
this should maybe be part of `tui::restore()`?

sayan-oai committed:
gate import to unix

nornagon-openai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1004,6 +1011,70 @@ impl App {
+            tracing::warn!("failed to restore terminal modes before editor: {err}");
+        }
+
+        let editor_result = editor::run_editor(&seed, &editor_cmd).await;
```
what does the terminal window look like while the editor is running?

sayan-oai replied:
it looks "the same" (just as if you tabbed away from the terminal, so the cursor isn't blinking).

if you tab back into the terminal while the `code` window is open and add text to the prompt, that text is retained (the resulting text from the `code` window is prepended to the text you added). but this seems like an edge case.

nornagon-openai replied:
hm... but didn't we disable all our terminal modes? we shouldn't be running the TUI at the same time as the editor is open, I think.

e.g. what happens if you end up in an odd state, e.g. an approval dialog is showing, or you're in transcript mode?

nornagon-openai replied:
i think this should behave roughly the same as `^Z`.

sayan-oai committed:
simplify errors, rm Option, disambiguate image placeholders

sayan-oai committed:
add external editor footer

sayan-oai committed:
comment + clean up

sayan-oai commented on codex-rs/tui/src/bottom_pane/chat_composer.rs:
```diff
@@ -281,6 +279,85 @@ impl ChatComposer {
+                occurrences.push((pos, *placeholder));
+            }
+        }
+        occurrences.sort_unstable_by_key(|(pos, _)| *pos);
```
added `#N` to disambiguate (`[screenshot.png 10x10]`, `[screenshot.png 10x10 #2]`)

sayan-oai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1004,6 +1011,70 @@ impl App {
+                // Empty file clears the composer.
+                self.chat_widget.apply_external_edit(String::new());
+                tui.frame_requester().schedule_frame();
+            }
```
good point, rm'd this extra case.

sayan-oai committed:
mv const to correct area

sayan-oai committed:
prevent shortcut while bottom pane in use

sayan-oai commented:
@codex review this

hasan-ozdemir commented:
Dear Participants; i have also implemented "ctrl+e:edit prompt in external editor" feature on windows since i was not aware of the improvement on this issue yet.
Now, it can open default notepad on windows.
Then how do i bring up my implementation in my topic branch so that you can check for the potential integration option into your solution if you like?
BTW: i have implemented it as a modular .JS extension along with an extension host on Rust TUI side.

sayan-oai committed:
fix leaky map; compute image placeholders on demand

sayan-oai committed:
fix terminal flushing

sayan-oai committed:
fix footer end calculation

sayan-oai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1004,6 +1018,180 @@ impl App {
+            return None;
+        }
+        Some(row)
+    }
```
thanks for catching! fixed.

sayan-oai commented on codex-rs/tui/src/bottom_pane/chat_composer.rs:
```diff
@@ -383,6 +463,19 @@ impl ChatComposer {
 
+    fn next_image_placeholder(&mut self, base: &str) -> String {
+        let counter = self
+            .image_placeholder_counters
```
Ahh, thanks. I changed it to calculate placeholders on demand instead of using the map.

sayan-oai committed:
whitespace

sayan-oai commented:
@codex review this

sayan-oai committed:
rework to set app state and avoid forced renders/cursor moves

sayan-oai committed:
keep raw mode enabled

sayan-oai committed:
rm cursor moving + clearing logic bc we stay in raw mode

sayan-oai committed:
clean up tests

sayan-oai committed:
cleanup

sayan-oai committed:
cleanup

nornagon-openai commented on codex-rs/tui/src/lib.rs:
```diff
@@ -44,6 +44,7 @@ mod clipboard_paste;
 mod color;
 pub mod custom_terminal;
 mod diff_render;
+mod editor;
```
nit: 
```suggestion
mod external_editor;
```

nornagon-openai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1099,6 +1127,132 @@ impl App {
+        tokio::spawn(async move {
+            time::sleep(Duration::from_millis(500)).await;
+            tx.send(AppEvent::ExternalEditorRequestTimeout);
+        });
```
what's the reasoning behind this timeout? why would we not get a draw in time? what behavior does this catch?

sayan-oai replied:
ah, this was leftover from when we paused polling on crossterm rather than dropping the `EventStream` entirely. rm'd since that was merged, thanks.

sayan-oai committed:
cleanup + reorganize

sayan-oai committed:
refactor terminal release + regain to single operation

sayan-oai commented on codex-rs/tui/src/app.rs:
```diff
@@ -1099,6 +1127,132 @@ impl App {
+        let restore_modes = tui::restore_keep_raw();
+        if let Err(err) = restore_modes {
+            tracing::warn!("failed to restore terminal modes before editor: {err}");
+        }
```
good point. refactored to this pattern in `tui.rs`.

nornagon-openai commented on codex-rs/tui/src/app.rs:
```diff
@@ -510,6 +515,10 @@ impl App {
     ) -> Result<bool> {
+        // If the external editor is active, don't process any tui events.
+        if self.chat_widget.external_editor_state() == ExternalEditorState::Active {
+            return Ok(true);
```
Does this ever actually run, since we're paused the tui event stream while the editor is active?

nornagon-openai commented on codex-rs/tui/src/tui.rs:
```diff
@@ -118,18 +119,82 @@ impl Command for DisableAlternateScroll {
+}
+
+#[cfg(unix)]
+pub(crate) fn flush_terminal_input_buffer() {
```
would love a doc comment on this function (and the windows variant)

nornagon-openai reviewed (approved):
Thanks for your patience and attention to detail, this is looking good! Some minor comments below but otherwise good to merge.

One minor thing for a followup: for non-terminal editors (e.g. `code --wait`), it would be nice to be able to cancel waiting for the editor to quit, in case it got stuck somehow. I'm not totally sure if this is possible to do given we're pausing the terminal event stream, but maybe we could still watch for ^C or something?

sayan-oai committed:
rework opening editor to release filehandle

windows doesnt allow saving files directly while another process has a handle, so rework editor logic to keep a temppath rather than the handle itself. this keeps windows happy and works with mac/linux

sayan-oai committed:
cleanup

sayan-oai committed:
handle editors on windows

sayan-oai committed:
use winsplit for windows parsing

sayan-oai committed:
Merge branch 'main' into dev/sayan/prompt-editor
